engine/errors.go

```diff
@@ -0,0 +1,51 @@
+package engine
+
+import "errors"
+
+var (
+	// ErrFeatureNotSupported is a common error that may be returned from optional
+	// Engine methods to indicate the engine implementation doesn't support the
+	// given feature.
+	//
+	// Note, all methods are allowed to return this error, some methods are
+	// required, and may not return this error.
+	//
+	// When the worker encounters this error from an optional method, it should
+	// workaround if possible, but most likely resolve the task as "exception"
+	// with reason "malformed-payload".
+	ErrFeatureNotSupported   = errors.New("Feature not support by current engine")
+	ErrResourceNotFound      = errors.New("The referenced resource wasn't found")
+	ErrSandboxTerminated     = errors.new("The Sandbox has terminated")
+	ErrSandboxAborted        = errors.new("Exection of sandbox was aborted")
+	ErrNonFatalInternalError = errors.new("Engine encountered a non-fatal internal error")
+	ErrContractViolation     = errors.new("Engine has detected a contract violation")
+	ErrEngineIsSingleton     = errors.New("Engine cannot run multiple sandboxes in parallel")
+	ErrEngineNotSupported    = errors.New("Engine is not available in the current configuration")
+	ErrEngineUnknown         = errors.New("Engine with the given doesn't exist")
+)
```

Comment:
So the case here is a `plugin` that calls a method on `engine` and gets `ErrFeatureNotSupported`.

If the plugin forwards this error to it's parent, task execution will abort and the worker will have an internal error. As `plugin` isn't allowed to return the  `ErrFeatureNotSupported`.

For each method in an interface I specify which non-fatal errors the implementor may return. If the implementor returns an error that isn't declared as non-fatal we won't attempt to handle it in any graceful manner. Most likely we'll just report the task as exception, writing some logs, and possibly crash the machine to prevent trashing of tasks (maybe there is a threshold of tasks you may thrash before crashing).

But essentially, if an error isn't listed as non-fatal the error won't be handled gracefully. So a plugin that just forwards `ErrFeatureNotSupported` has an internal error. It may convert `ErrFeatureNotSupported` into a `MalformedPayloadError` saying that the proxy or cache attach option is invalid...